### PR TITLE
Core/Entities : Conversation

### LINKED ISSUE #17992: Core/Entity : new dialogs in WoW Legion
**Description**:

Hello, I am trying to recreate the initial mission of the hunters of demons, someone has some idea of since it could do http://oi68.tinypic.com/x5umtc.jpg
What in warcraft the IIIrd was a dialog, now my question is, for this type of dialogs it is a necessary algun script in lua or in c ++

**Current behaviour**: not work

**Expected behaviour**: Tell us what should happen instead

--removed--

**Steps to reproduce the problem**:

1.It is necessary opcodes, or is just a matter of scrips

**Branch(es)**: 3.3.5 / 6.x (Select the branch(es) affected by this issue)

**Tags 7.0.3/22594**

Edited by DDuarte

### PR DESCRIPTION
**Pull request by Traesh**
**Changes proposed:**
-  Implement Conversation Entity, used to display ingame dialog
-  Command .debug conversation # convId
-  Command .Reload conversation_template
- Add 5 tables for templates, actors & lines

**Target branch(es):** 6.x

**Issues addressed:** Closes #17992 ([Core/Entity : new dialogs in WoW Legion](https://github.com/TrinityCore/TrinityCore/issues/17992))

**Tests performed:** (Does it build, tested in-game, etc)
Built
Tested ingame with conversation id 705    (.debug conv 705)
Tested ingame with conversation id 1025  (.debug conv 1025)

**Known issues and TODO list:**
- None ATM

(Re-Opening PR due to a fail on previous repo, all reviews have been fixed)

---

(minor edit/addendum by tkrokli)

### CODE REVIEW

**Comment by Pixel-Pirate** [comment]
Your pr is interesting but I think his is more convenient to make it work with the hotfixes databse

**Comment by Traesh** [comment]
Hello :) What do you mean by making it work with the hotfixes database ?

File: src/server/game/Entities/Conversation/Conversation.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    if (IsInWorld())
+    {
+        SendObjectDeSpawnAnim(GetGUID());
+        RemoveFromWorld();
+        AddObjectToRemoveList();
+    }
+}
+
+
+bool Conversation::CreateConversation(ObjectGuid::LowType guidlow, uint32 conversationEntry, Unit* caster, SpellInfo const* spell, Position const& pos)
```
Comment (tkrokli) [inline src/server/game/Entities/Conversation/Conversation.cpp]:
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Entities/Conversation/Conversation.cpp:87:125: fatal error: 
      unused parameter 'spell' [-Wunused-parameter]
  ...uint32 conversationEntry, Unit* caster, SpellInfo const* spell, Positio...
                                                              ^
```

File: src/server/game/Spells/SpellEffects.cpp [hunk 2]
```diff
@@ -5681,6 +5682,24 @@ void Spell::EffectCreateGarrison(SpellEffIndex effIndex)
     unitTarget->ToPlayer()->CreateGarrison(GetEffect(effIndex)->MiscValue);
 }
 
+void Spell::EffectCreateConversation(SpellEffIndex effIndex)
```
Comment (tkrokli) [inline src/server/game/Spells/SpellEffects.cpp]:
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Spells/SpellEffects.cpp:5685:52: fatal error: 
      unused parameter 'effIndex' [-Wunused-parameter]
void Spell::EffectCreateConversation(SpellEffIndex effIndex)
                                                   ^
```

**Comment by tkrokli** [comment]
Good to see that there is only the 'auth' database left to fix (ToDo when this PR gets merged).
Anything else left to check or review in this PR? :)

**Comment by Druson** [comment]
seems ConversationLine.db2 contain the next conversation  line ID of current conversation line.  So can table conversation_lines just contain the conversation ID and first line ID only? 

And the field Unk1 in the table conversation_line_template should be the duration of previous conversation_line. The duration should be different for different locale.

**Comment by funjoker** [comment]
News?

**Comment by Aokromes** [comment]
continuous-integration/travis-ci/pr — The Travis CI build failed

**Comment by funjoker** [comment]
Travis fail due to SQL?

**Comment by tkrokli** [comment]
Yes, it is only because the auth database update list is not adjusted yet.
```
- "sql/updates/auth/9999_99_99_18010_auth.sql" is missing in the 'auth'.'updates' table.
Fatal error:

  The Database Updater is broken for the 'auth' database,
  because the applied update is missing in the 'auth'.'updates' table.

How to fix:
  Insert the missing names of the already applied sql updates
  to the 'updates' table of the 'auth' base dump ('sql/base/auth_database.sql').
```
Then again, even if the update is added in the PR, it needs to be redone after merge.
So it won't make much difference if it is included in the PR, mostly only to the Travis result.

**Comment by funjoker** [comment]
Merge?

**Comment by tkrokli** [comment]
This merge needs to be done by someone with experience in creating the correct hash change to `auth`.`updates` in addition to the normal renaming of the SQL files in this PR.

File: src/server/game/Spells/SpellEffects.cpp [hunk 3]
```diff
@@ -5681,6 +5682,24 @@ void Spell::EffectCreateGarrison(SpellEffIndex effIndex)
...
+
+    Position pos;
+    if (!m_targets.HasDst())
+        pos = GetCaster()->GetPosition();
+    else
+        pos = destTarget->GetPosition();
+
+    uint32 conversationEntry = effectInfo->MiscValue;
+
+    Conversation* conversation = new Conversation;
```
Comment (Naios) [inline src/server/game/Spells/SpellEffects.cpp]: Use snart pointers please

File: src/server/game/Globals/ObjectMgr.h [hunk 4]
```diff
@@ -1424,6 +1430,15 @@ class TC_GAME_API ObjectMgr
             return nullptr;
         }
 
+        ConversationTemplate const* GetConversationTemplate(uint32 id) const
+        {
+            auto itr = _conversationTemplateStore.find(id);
+            if (itr != _conversationTemplateStore.end())
+                return &itr->second;
+
+            return nullptr;
```
Comment (Naios) [inline src/server/game/Globals/ObjectMgr.h]: You could use `Optional` here to clarify that your pointer might be null.

File: src/server/game/Spells/SpellEffects.cpp (hunk 3 above)
Comment (Shauren) [inline src/server/game/Spells/SpellEffects.cpp]: and noooo

File: src/server/game/Globals/ObjectMgr.h (hunk 4 above)
Comment (Shauren) [inline src/server/game/Globals/ObjectMgr.h]: noooo

File: src/server/game/Entities/Conversation/Conversation.cpp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        Remove(); // expired
+
+    WorldObject::Update(p_time);
+}
+
+void Conversation::Remove()
+{
+    if (IsInWorld())
+    {
+        SendObjectDeSpawnAnim(GetGUID());
```
Comment (joschiwald) [inline src/server/game/Entities/Conversation/Conversation.cpp]: it's a gameobject only packet

**Comment by joschiwald** [comment]
what is the current state of this pr?

**Comment by funjoker** [comment]
well he is active thing he finishes areatrigger first

**Comment by tkrokli** [comment]
BTW, in your next commit, please make sure that you change the location of your auth sql update:
`sql/updates/auth/6.x/9999_99_99_18010_auth.sql` needs to be moved to
`sql/updates/auth/master/9999_99_99_18010_auth.sql`

**Comment by matukaa** [comment]
Umm, up? :D

**Comment by tkrokli** [comment]
Going to push another commit in this PR branch, moving the `[ ]_auth.sql` from `6.x` to `master`.

File: sql/updates/auth/6.x/9999_99_99_18010_auth.sql [hunk 6]
```diff
@@ -0,0 +1,7 @@
+DELETE FROM `rbac_permissions` WHERE `id` IN (851, 852);
+INSERT INTO `rbac_permissions` (`id`, `name`) VALUES
+(851, 'Command: .reload conversation_template'),
+(852, 'Command: .debug conversation');
```
Comment (joschiwald) [inline sql/updates/auth/6.x/9999_99_99_18010_auth.sql]: wrong ids

Comment (tkrokli) [inline sql/updates/auth/6.x/9999_99_99_18010_auth.sql]: Thanks, I will fix that too. Forgot to change it outside of solving the conflict.

File: sql/updates/hotfixes/master/9999_99_99_18010_hotfixes.sql [hunk 7]
```diff
@@ -0,0 +1,14 @@
+DROP TABLE IF EXISTS `conversation_line`;
```
Comment (tkrokli) [inline sql/updates/hotfixes/master/9999_99_99_18010_hotfixes.sql]: This `DROP TABLE` command does not match `CREATE TABLE`.
Going to remedy this in my commit as well.

File: src/server/database/Database/Implementation/HotfixDatabase.cpp [hunk 8]
```diff
@@ -175,6 +175,10 @@ void HotfixDatabaseConnection::DoPrepareStatements()
     PrepareStatement(HOTFIX_SEL_CINEMATIC_SEQUENCES, "SELECT ID, SoundID, Camera1, Camera2, Camera3, Camera4, Camera5, Camera6, Camera7, Camera8"
         " FROM cinematic_sequences ORDER BY ID DESC", CONNECTION_SYNCH);
 
+    // ConversationLine.db2
+    PrepareStatement(HOTFIX_SEL_CONVERSATION_LINE, "SELECT ID, Unk1, Unk2, Unk3, NextLineID, Unk4, UnkBoolean, Unk5, Unk6"
+        " FROM conversation_line ORDER BY ID DESC", CONNECTION_SYNCH);
```
Comment (tkrokli) [inline src/server/database/Database/Implementation/HotfixDatabase.cpp]:
```
In mysql_stmt_prepare() id: 48, sql: "SELECT ID, Unk1, Unk2, Unk3, NextLineID, Unk4, UnkBoolean, Unk5, Unk6 FROM conversation_line ORDER BY ID DESC"
Table 'hotfixes.conversation_line' doesn't exist
```
I presume the intended table name is `conversation_line_template`.
Adding this fix to my commit too.

Comment (tkrokli) [inline src/server/database/Database/Implementation/HotfixDatabase.cpp]: Sorry, I was wrong. The correct table name here is `conversation_template`.
Fix rework still in progress.

**Comment by joschiwald** [comment]
Don't touch hotfix database

**Comment by tkrokli** [comment]
I only renamed the table name `conversation_line` to `conversation_template`, which was already referenced in the already existing files `sql/updates/hotfixes/master/2017_99_99_18010_hotfixes.sql` and `src/server/database/Database/Implementation/HotfixDatabase.cpp`. Maybe the error message I posted yesterday was a bit confusing or misleading?

**Comment by Shauren** [comment]
Thats where you are wrong. hotfix tables are autogenerated and must strictly follow naming of its corresponding .db2 file

**Comment by tkrokli** [comment]
As you wish, reverting my changes related to Hotfixes.

**Comment by tkrokli** [comment]
I don't get it. The PR compiles, but my local worldserver won't complete starting up. From Server.log:
```
Could not prepare statements of the Hotfix database, see log for details.
```
From DBErrors.log:
```
In mysql_stmt_prepare() id: 48, sql: "SELECT ID, Unk1, Unk2, Unk3, NextLineID, Unk4, UnkBoolean, Unk5, Unk6 FROM conversation_line ORDER BY ID DESC"
Table 'hotfixes.conversation_line' doesn't exist
```
This happens while starting worldserver, rebuilding all the DBs from scratch (removed all DBs before start).

**Comment by Shauren** [comment]
Look closely at that sql file
```
DROP TABLE IF EXISTS `conversation_line`;
CREATE TABLE `conversation_template` (
```

**Comment by tkrokli** [comment]
Sure, that is one of the 2 things I corrected in my previous commit, but I was asked to revert it.
Maybe I am too slow (or braindead) to see what the correct solution is. Please enlighten me.

**Comment by Shauren** [comment]
Change the second line only to be `conversation_line`, not the first line and core

**Comment by tkrokli** [comment]
Any other changes required in this PR before merging?

File: sql/updates/hotfixes/master/2017_99_99_18010_hotfixes.sql [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+CREATE TABLE `conversation_line` (
+  `Id` int (10) UNSIGNED NOT NULL,
+  `Unk1` int (10) UNSIGNED NOT NULL,
+  `Unk2` int (10) UNSIGNED NOT NULL,
+  `Unk3` int (10) UNSIGNED NOT NULL,
+  `NextLineID` smallint (5) UNSIGNED NOT NULL,
+  `Unk4` smallint (5) UNSIGNED NOT NULL,
+  `UnkBoolean` tinyint (3) UNSIGNED NOT NULL,
+  `Unk5` tinyint (3) UNSIGNED NOT NULL,
+  `Unk6` tinyint (3) UNSIGNED NOT NULL,
```
Comment (joschiwald) [inline sql/updates/hotfixes/master/2017_99_99_18010_hotfixes.sql]: one of these unknowns is a string field, but the dbcviewer does not show any useful things

File: src/server/game/Globals/ObjectMgr.cpp [hunk 10]
```diff
@@ -9570,3 +9570,108 @@ void ObjectMgr::LoadSceneTemplates()
...
+
+            QueryResult actors = WorldDatabase.PQuery("SELECT ConversationActorId FROM conversation_actors WHERE ConversationId = %u ORDER BY Idx", conversationTemplate.Id);
+
+            if (actors)
+            {
+                do
+                {
+                    Field* fields = actors->Fetch();
+                    uint32 actorId = fields[0].GetUInt32();
+                    ConversationActorTemplate conversationActorTemplate = _conversationActorTemplateStore[actorId];
```
Comment (joschiwald) [inline src/server/game/Globals/ObjectMgr.cpp]: ConversationActorTemplate&

File: src/server/game/Globals/ObjectMgr.cpp [hunk 11]
```diff
@@ -9570,3 +9570,108 @@ void ObjectMgr::LoadSceneTemplates()
...
+
+                    conversationTemplate.Actors.push_back(conversationActorTemplate);
+                } while (actors->NextRow());
+            }
+
+            ConversationLineEntry const* currentConversationLine = sConversationLineStore.LookupEntry(conversationTemplate.FirstLineId);
+
+            while (currentConversationLine != nullptr)
+            {
+                ConversationLineTemplate conversationActorTemplate = _conversationLineTemplateStore[currentConversationLine->ID];
```
Comment (joschiwald) [inline src/server/game/Globals/ObjectMgr.cpp]: ref

File: src/server/game/Entities/Conversation/Conversation.cpp [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+    ConversationTemplate const* conversationTemplate = sObjectMgr->GetConversationTemplate(conversationEntry);
+
+    if (!conversationTemplate)
+        return false;
+
+    SetUInt32Value(CONVERSATION_FIELD_LAST_LINE_DURATION, conversationTemplate->LastLineDuration);
+    _duration = conversationTemplate->LastLineDuration;
+
+    uint16 actorsIndex = 0;
+    for (ConversationActorTemplate actor : conversationTemplate->Actors)
```
Comment (joschiwald) [inline src/server/game/Entities/Conversation/Conversation.cpp]: const&

File: src/server/game/Entities/Conversation/Conversation.cpp [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+        SetDynamicStructuredValue(CONVERSATION_DYNAMIC_FIELD_ACTORS, actorsIndex++, &actor);
+
+    uint16 linesIndex = 0;
+    for (ConversationLineTemplate line : conversationTemplate->Lines)
```
Comment (joschiwald) [inline src/server/game/Entities/Conversation/Conversation.cpp]: const&

File: sql/updates/hotfixes/master/2017_99_99_18010_hotfixes.sql (hunk 9 above)
Comment (joschiwald) [inline sql/updates/hotfixes/master/2017_99_99_18010_hotfixes.sql]: @Shauren: any idea what are these unknowns?

**Comment by funjoker** [comment]
Any news about it ? :D

File: sql/updates/hotfixes/master/2017_99_99_18010_hotfixes.sql (hunk 9 above)
Comment (funjoker) [inline sql/updates/hotfixes/master/2017_99_99_18010_hotfixes.sql]: no don't think there is a string inside

Comment (funjoker) [inline sql/updates/hotfixes/master/2017_99_99_18010_hotfixes.sql]: Nope no strings.
All strings are inside the broadcast_text at hotfixes db.
Unk1 is the broadcast_text.Id !

**Comment by funjoker** [comment]
```
DROP TABLE IF EXISTS `conversation_line`;
CREATE TABLE `conversation_line` (
  `Id` int (10) UNSIGNED NOT NULL,
  `Unk1` int (10) UNSIGNED NOT NULL,
  `Unk2` int (10) UNSIGNED NOT NULL,
  `Unk3` int (10) UNSIGNED NOT NULL,
  `NextLineID` smallint (5) UNSIGNED NOT NULL,
  `Unk4` smallint (5) UNSIGNED NOT NULL,
  `UnkBoolean` tinyint (3) UNSIGNED NOT NULL,
  `Unk5` tinyint (3) UNSIGNED NOT NULL,
  `Unk6` tinyint (3) UNSIGNED NOT NULL,
  `VerifiedBuild` int (10) UNSIGNED NOT NULL DEFAULT 0,
  PRIMARY KEY (`Id`)
) ENGINE=MyISAM DEFAULT CHARSET=utf8;
```

Unk1 is the broadcast_text.Id
So is the db2 too

**Comment by funjoker** [comment]
Unk3 is a break between the running and following Text in milliseconds

File: sql/updates/hotfixes/master/2017_99_99_18010_hotfixes.sql (hunk 9 above)
Comment (funjoker) [inline sql/updates/hotfixes/master/2017_99_99_18010_hotfixes.sql]: Unk3 is a break between the running and following Text in milliseconds

**Comment by tkrokli** [comment]
OK, so maybe we could borrow the same naming style as for the `broadcast_text.ID` in the `creature_text` table, which is `BroadcastTextId` as a name for the column `Unk1`? (Just a suggestion.)

I am more uncertain about how to pick a good name to replace `Unk3`, because I don't remember an existing column using that kind of information (but I may have overlooked something). I presume the name could usefully contain a 3-part naming reference to
1. conversation or text
2. break or interval
3. ms (milliseconds) or time

Some possible suggestions could be `text_interval_ms` or `ConversationBreakTime`. Maybe there are naming conventions to be followed, I don't remember at the moment. Not sure if it is practical to add that change in this PR, though.

**Comment by funjoker** [comment]
I think ConversationBreakTime might be bit long but good.

**Comment by tkrokli** [comment]
OK, it was just a suggestion. Shorter suggestions could be `TextBreakMS` / `LineIntervalMS` and so on.

**Comment by offl** [comment]
```sql
  `NextLineTime` int (10) UNSIGNED NOT NULL,
  `NextLineID` smallint (5) UNSIGNED NOT NULL,
```

**Comment by funjoker** [comment]
> ```
>  `NextLineTime` int (10) UNSIGNED NOT NULL,
>   `NextLineID` smallint (5) UNSIGNED NOT NULL,
> ```

Could maybe be confusing about the thing the column is for

**Review by funjoker — APPROVED** [review]

**Comment by funjoker** [comment]
One of the Unk in conversation_line is the type of the conversation, but donno which =/

**Comment by funjoker** [comment]
```
 struct ConversationLineEntry
 {
     uint32 ID;
     uint32 BroadcastTextId;
     uint32 Unk2;
     uint32 LineIntervalMS;
     uint16 NextLineID;
     uint16 Unk4; // Flag ?
     uint8 UnkBoolean;
     uint8 Unk5;
     uint8 Unk6;
 };
```

Unk2; <--- SpellVisualKitID

**Comment by funjoker** [comment]
```
struct ConversationLineTemplate
 {
     uint32 Id;      // Link to ConversationLine.db2
     uint32 PreviousLineDuration;
     uint32 Unk2;
     uint32 Unk3;
 };
```

`Unk2` is it maybe AnimKitId ?
Because the avatars are sometimes using angry movement.

**Comment by funjoker** [comment]
SPELL_EFFECT_219  -> SPELL_EFFECT_CONVERSATION 
Not proofed :D at work atm so someone might take a look ^^

**Comment by Traesh** [comment]
@funjoker It's already in this PR :p

**Comment by funjoker** [comment]
OKAY :D sorry ^^

**Comment by Aokromes** [comment]
Conflicting files
sql/base/auth_database.sql
src/server/game/Accounts/RBAC.h

**Review by ghost — APPROVED** [review]

File: src/server/game/DataStores/DB2Stores.cpp [hunk 14]
```diff
@@ -58,6 +58,7 @@ DB2Storage<ChrRacesEntry>                       sChrRacesStore("ChrRaces.db2", C
 DB2Storage<ChrSpecializationEntry>              sChrSpecializationStore("ChrSpecialization.db2", ChrSpecializationLoadInfo::Instance());
 DB2Storage<CinematicCameraEntry>                sCinematicCameraStore("CinematicCamera.db2", CinematicCameraLoadInfo::Instance());
 DB2Storage<CinematicSequencesEntry>             sCinematicSequencesStore("CinematicSequences.db2", CinematicSequencesLoadInfo::Instance());
+DB2Storage<ConversationLineEntry>               sConversationLineStore("ConversationLine.db2", ConversationLineLoadInfo::Instance());
```
Comment (tkrokli) [inline src/server/game/DataStores/DB2Stores.cpp]:
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/DataStores/DB2Stores.cpp:61:49: fatal error: 
      no matching constructor for initialization of
      'DB2Storage<ConversationLineEntry>'
  ...sConversationLineStore("ConversationLine.db2", ConversationLineLoadInfo::Instance());
     ^                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment (Shauren) [inline src/server/game/DataStores/DB2Stores.cpp]: Yeah, I changed the loadinfo part when making db2 loaders usable by extractors, PR author needs to rerun latest https://github.com/Shauren/WowClientDB2MySQLTableGenerator

**Comment by wonder2011** [comment]
Merge?

File: src/server/database/Database/Implementation/HotfixDatabase.cpp [hunk 15]
```diff
@@ -180,7 +180,7 @@ void HotfixDatabaseConnection::DoPrepareStatements()
         " FROM cinematic_sequences ORDER BY ID DESC", CONNECTION_SYNCH);
 
     // ConversationLine.db2
-    PrepareStatement(HOTFIX_SEL_CONVERSATION_LINE, "SELECT ID, BroadcastTextId, Unk2, LineIntervalMS, NextLineID, Unk4, UnkBoolean, Unk5, Unk6"
+    PrepareStatement(HOTFIX_SEL_CONVERSATION_LINE, "SELECT ID, BroadcastTextId, BroadcastTextId, LineIntervalMS, NextLineID, Unk4, UnkBoolean, Unk5, Unk6"
```
Comment (Pixel-Pirate) [inline src/server/database/Database/Implementation/HotfixDatabase.cpp]: Not BroadcastTextId but SpellVisualKitID ;)

File: src/server/game/Entities/Conversation/Conversation.cpp [hunk 16]
```diff
@@ -102,7 +102,7 @@ bool Conversation::CreateConversation(ObjectGuid::LowType guidlow, uint32 conver
     if (!conversationTemplate)
         return false;
 
-    SetUInt32Value(CONVERSATION_FIELD_LAST_LINE_DURATION, conversationTemplate->LastLineDuration);
+    SetUInt32Value(CONVERSATION_LAST_LINE_END_TIME, conversationTemplate->LastLineDuration);
```
Comment (Shauren) [inline src/server/game/Entities/Conversation/Conversation.cpp]: well, are you sure about this now?

File: src/server/game/Entities/Conversation/Conversation.h [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+#include "Object.h"
+#include "ObjectAccessor.h"
+
+class Unit;
+class SpellInfo;
+
+struct ConversationActorTemplate
+{
+    uint32 Id;
```
Comment (joschiwald) [inline src/server/game/Entities/Conversation/Conversation.h]: the first 4 uint can be a full guid

File: src/server/game/Globals/ConversationDataStore.cpp [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+            conversationTemplate.LastLineEndTime    = fields[2].GetUInt32();
+
+            if (QueryResult actors = WorldDatabase.PQuery("SELECT ConversationActorId FROM conversation_actors WHERE ConversationId = %u ORDER BY Idx", conversationTemplate.Id))
+            {
+                do
+                {
+                    Field* fields = actors->Fetch();
+                    uint32 actorId = fields[0].GetUInt32();
+                    if (ConversationActorTemplate const* conversationActorTemplate = Trinity::Containers::MapGetValuePtr(_conversationActorTemplateStore, actorId))
+                        conversationTemplate.Actors.push_back(*conversationActorTemplate);
```
Comment (Shauren) [inline src/server/game/Globals/ConversationDataStore.cpp]: Whats the logic behind just copying the actor inside the conversation template?

Comment (joschiwald) [inline src/server/game/Globals/ConversationDataStore.cpp]: i don't see it too, i just copied stuff from ObjectMgr with few modifications, i'm going to change something about saving actors after i have move infos about them, do you have some? about the unknowns?

File: src/server/game/Globals/ConversationDataStore.cpp [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ConversationDataStore.h"
+#include "Containers.h"
+#include "Conversation.h"
```
Comment (Shauren) [inline src/server/game/Globals/ConversationDataStore.cpp]: This goes against my idea of cleaning up headers - it should be `ConversationDataStore.h` that defines the template structures (or even a separate header), not `Conversation.h` (heavy header to include due to Object.h - pulls all of map & grid)

**Comment by joschiwald** [comment]
i'm going to merge this pr tomorrow

**Pull request merged**